app/lib/api_error/saved_page_not_persisted.rb

```diff
@@ -0,0 +1,21 @@
+module ApiError
+  class SavedPageNotPersisted < ApiError::Base
+    def initialize(page_path:)
+      @page_path = page_path
+    end
+
+    def status_code
+      :bad_request
```

Comment:
Should this be a 422?

Also, this class is pretty vague.  Can the validation errors be added to the `extra_detail`?

Reply:
Yeah it is pretty vauge, it fell into a similar category as the other destroy method.
This one was initially what was returned once as a final else to a `rescue` if we'd already rescued all known errors (ones with specific messages), and yet the resource remained not `.persisted?`

However since switching these to use ActiveModel validations it gets a bit more compliacted.
We want to use `save` on the model, as we can then capture `:blank` and return a more meaingful error name (if we throw an error I need to return to using rescue as flow control).

However there may be errors that I have not handled for, in that case I'd wanted to a generic error that I throw as a last ditch effort. Again so if these pop up in sentry we can act on them quickly, but also to cater for any scenario i've not thought about.

I considered if these were 5xx errors (our fault) or 4xx (theirs), and went with bad request because I reckon to end up here there must have been a wacky request that:
- was authenticated
- had a non blank, valid page path

But yet could not be persisted...

Reply:
Ok, in that case, rather than having a separate errors for some validation errors, why not have one error type for all validation errors?

```ruby
raise ApiError::CannotSavePage, { page_path: params[:page_path], errors: saved_page.errors }
```

I think a useful thing to keep in mind for introducing new API errors is whether splitting an error into more specific types lets the API client react in a way it didn't have enough information for before.  Is an API client really going to handle `SavedPageDuplicate`, `SavedPagePathRequired`, and `SavedPageNotPersisted` differently?

Reply:
Probably not, that's a good point. So `create` would just be
```ruby
  # POST /api/saved_pages
  def create
    saved_page = SavedPage.new(oidc_user_id: user.id, page_path: params[:page_path])

    if saved_page.save
      render_api_response(saved_page: saved_page.to_hash)
    else
      raise ApiError::CannotSavePage, { page_path: params[:page_path], errors: saved_page.errors }
    end
  end
```

Reply:
Humm, One unified error message sounds good, but I'm also a bit nervous of sending raw errors back out over HTTP?

Stop me if htis is too paranoid?
But i never quite know the full range of what we might expose there, wouldn't want some innards of the app to give any good intellegence to an attacker? Or at least one who hadn't found the source code here yet...

Reply:
I guess what'd i'd been trying to do is capture specific error states where I thought "there's a likelyhood this could crop up" and return specific, actionable error messages.

Then at the bottom have an `else` category, where something has gone wrong, but I didn't forsee what it was here.
If we ever saw that in sentry it'd be a guide for devs to go back, investigate and likely add another test case + more specific error, or fix a bug

Reply:
We're not giving away any internal details of the app because we're not directly exposing account-api publicly.  And also returning validation errors in a 422 response is a pretty common pattern for our other API apps (eg, publishing-api does it).